### PR TITLE
Support differentiation across collective APIs.

### PR DESCRIPTION
**Pull request by dlibenzi**

### CODE REVIEW

**Comment by ronghanghu** [comment]
Thanks for the implementation! @dlibenzi I have a question on this PR: do you also need to do an all_reduce operation in the backward pass?

in FAIR's VISSL package, a differentiable `GatherLayer` is implemented as follows https://github.com/facebookresearch/vissl/blob/master/vissl/utils/distributed_gradients.py#L15-L31. The reduce op is also used in the backward pass.

If I understand correctly, the gradients should also be reduced across devices. Take all_reduce as an example, if on device 1 we have y1 = x1 + x2 and on device 2 we have y2 = x1 + x2, then in the backward pass we should have dx1 = dy1 + dy2 and dx2 = dy1 + dy2. But the current implementation only has dx1 = dy1 and dx2 = dy2.

Could you take a look? Thanks!

**Comment by ailzhang** [comment]
@ronghanghu Thanks for reaching out! Yea I think you are right about we need paired reduce/gather in forward/backward. Created an issue in https://github.com/pytorch/xla/issues/2989. If you're interested in contributing/reviewing the fix please let us know :D 
cc: @davidel just in case I'm missing something.

**Comment by davidel** [comment]
AFAICT, if we do a _reduce(SUM)_, we have:
```
S(x) = F1(x) + ... + Fn(x)
```
Where _Fi(x)_ are the per-device functions/graphs (being same or different does not matter), which are north of the reduce operation.
Then, downstream, we have the loss as _L(x)_, which is a function of some form of _S(x)_.
We enter the _all_reduce()_ backward with _dL(x)/dS(x)_ (using derivative notation _d_).
Every device path should get a gradient as _dL(x)/dFi(x)_ (_i_ in 1..n):
```
dL(x)/dFi(x) = dL(x)/dS(x) * dS(x)/dFi(x)
```
Where _dL(x)/dS(x)_ is the input gradient of the _all_reduce()_ backward, and _dS(x)/dFi(x)_ being 1 by _S(x)_ definition.

The question is whether the _dL(x)/dS(x)_ should be reduced (summed) across devices before propagating, and I believe yes, they should.

Care should be taken for existing users, as they will now have to step back their LR schedule (upstream gradients will be bigger).

**Comment by ronghanghu** [comment]
> The question is whether the dL(x)/dS(x) should be reduced (summed) across devices before propagating, and I believe yes, they should.

Thanks for the reply! One use case for this all-reduce with back-propagation is to implement a contrastive loss (such as in [CLIP](https://openai.com/blog/clip/)) across multiple devices. Summing the gradients _dL(x)/dS(x)_ makes the training consistent across different TPU core numbers, so that one can the same results under the a certain total (global) batch size whether using v3-128 or v3-256 TPUs.
